### ISSUE TITLE
Where were the classes (101, 51) declared?

### ISSUE THREAD

**Issue by EileenSchreiber** [bryanyzhu/Hidden-Two-Stream#12]
Hello,
I've been trying to find all occurrences, where the class number was declared. So far I found it in `end_train_val.prototxt`, `stack_motionnet_vgg16_deploy.prototxt` and `demo_hidden.py`. I would like to train the net on my own data, with less classes. Is that even possible? Where else do I have to change the class number? I still want to use your pretrained net with similar videos.

Thanks so much in advance.

**Comment by bryanyzhu** [comment]
Hi, it is definitely possible. And yes, you already found all the occurrences. There is nothing more to change. Of course, you need to change the input files (train_rgb_split1.txt, val_rgb_split1.txt), I assume you already did it. 

One thing to note is that, if you have a lot of training videos. It is good, you don't need to change anything, you just change the 'class_num' and start training. But if you have few training videos (say less than 5K), I suggest you reduce the learning rate, or even frozen the convolutional filter weights. Several epochs of fine tuning will be enough. The convergence should be fast.

**Comment by EileenSchreiber** [comment]
Hi,
after a couple of changes it is working. One question though, how does `late_fusion.m` get the files it needs? How did you compute them? (`spatial_quality100_split01.mat`, `temporal_hidden_split01.mat`, `temporal_tvl1_split01.mat`)

Thanks again.

**Comment by bryanyzhu** [comment]
You can check [here](https://github.com/bryanyzhu/Hidden-Two-Stream/blob/master/models/ucf101_split1_unsup_end/eval_ucf101/demo_hidden.py#L79-L83) for how to compute 'temporal_hidden_split01.mat'. The other two are similar. 

Basically, you get one 101X3783 matrix from spatial stream, and another 101X3783 matrix from temporal stream. You save them to disk using 'sio.savemat'. Then you load them, and do a weighted averaging of these two matrix, which is the late fusion. In the end, you use argmax to obtain the predicted action labels, which gives you a 3783X1 vector. Here, 101 is the number of classes in UCF101, 3783 is the number of videos in the test set of first split. Hope this is clear.

**Comment by EileenSchreiber** [comment]
Thank you.
But I'm still having trouble understanding it completely. So the `split1_hidden_end_before.mat` is the `temporal_hidden_split01.mat`, but what is the `split1_hidden_end.mat` then? And where do you set the matrix to 101X3783? I can't seem to find it. Because I have to change this then too, right?

Thanks in advance.

**Comment by bryanyzhu** [comment]
Oh, 'split1_hidden_end_before.mat' stores the 101-dim vectors before softmax operator, and 'split1_hidden_end.mat' stores the 101-dim vectors after softmax operator. Usually 'split1_hidden_end_before.mat' gives you better result. 

And I define the matrix [here](https://github.com/bryanyzhu/Hidden-Two-Stream/blob/master/models/ucf101_split1_unsup_end/eval_ucf101/demo_hidden.py#L37-L43). Specifically 
`predict_results_before = np.zeros(shape=dims, dtype=np.float64)`
Here, 'dims' is set as
`dims = (len(val_list), num_categories)`
val_list is how many test videos you have, which is 3783. num_categories is how many action classes you have, which is 101. You can change here accordingly.
